base/compiler/ssair/irinterp.jl

```diff
@@ -0,0 +1,400 @@
+        world = get_world_counter(interp)
+        value = try
+            Core._call_in_world_total(world, args...)
+        catch
+            return Union{}
+        end
+        if is_inlineable_constant(value) || call_result_unused(sv)
+            # If the constant is not inlineable, still do the const-prop, since the
+            # code that led to the creation of the Const may be inlineable in the same
+            # circumstance and may be optimizable.
+            return Const(value)
+        end
+    else
+        ir′ = codeinst_to_ir(interp, code)
+        if ir′ !== nothing
+            return ir_abstract_constant_propagation(interp, mi_cache, sv, mi, ir′, argtypes)
+        end
+    end
+    return nothing
+end
+
+function reprocess_instruction!(interp::AbstractInterpreter, ir::IRCode, mi::MethodInstance,
+                                mi_cache, sv::InferenceState,
+                                tpdum::TwoPhaseDefUseMap, idx::Int, bb::Union{Int, Nothing},
+                                @nospecialize(inst), @nospecialize(typ),
+                                phi_revisit)
+    function update_phi!(from, to)
+        if length(ir.cfg.blocks[to].preds) == 0
+            return
+        end
+        for idx in ir.cfg.blocks[to].stmts
+            stmt = ir.stmts[idx][:inst]
+            isa(stmt, Nothing) && continue
+            isa(stmt, PhiNode) || break
+            for (i, edge) in enumerate(stmt.edges)
+                if edge == from
+                    deleteat!(stmt.edges, i)
+                    deleteat!(stmt.values, i)
+                    push!(phi_revisit, idx)
+                    break
+                end
+            end
+        end
+    end
+
+    if isa(inst, GotoIfNot)
+        cond = argextype(inst.cond, ir)
+        if isa(cond, Const)
+            if isa(inst.cond, SSAValue)
+                kill_def_use!(tpdum, inst.cond, idx)
+            end
+            if bb === nothing
+                bb = block_for_inst(ir, idx)
+            end
+            if (cond.val)::Bool
+                ir.stmts[idx][:inst] = nothing
+                kill_edge!(ir, bb, inst.dest, update_phi!)
+            else
+                ir.stmts[idx][:inst] = GotoNode(inst.dest)
+                kill_edge!(ir, bb, bb+1, update_phi!)
+            end
+            return true
+        end
+        return false
+    else
+        if isa(inst, Expr) || isa(inst, PhiNode)
+            if isa(inst, PhiNode) || inst.head === :call || inst.head === :foreigncall || inst.head === :new
+                if isa(inst, PhiNode)
+                    rt = abstract_eval_phi(interp, inst, nothing, ir)
+                else
+                    (;rt, effects) = abstract_eval_statement_expr(interp, inst, nothing, ir, mi)
+                    # All other effects already guaranteed effect free by construction
+                    if is_nothrow(effects)
+                        ir.stmts[idx][:flag] |= IR_FLAG_EFFECT_FREE
+                    end
```

Comment:
Why do we want to handle this here rather than letting the optimizer to annotate this as like the regular absint->optimization chain?

Reply:
I'd just like to generally move towards inference annotating this information if it has it. We're already doing all the analysis, so doing it over again in the optimizer is wasteful. That said, this branch is dead anyway, since we're currently requiring nothrow, but I'll plan to address that after merge.